### ISSUE TITLE
Linux - when Upgrading from version 4.3.3.952 to 4.3.4.1084  the client always raises the exception «None of the discovered or specified addresses match the socket address family»

### ISSUE THREAD

**Issue by daframax** [dotnet/MQTTnet#1997]
### Describe the bug
Using the latest version on Linux (Mono) the Client always raises the *MqttCommunicationException* with the message: **None of the discovered or specified addresses match the socket address family** when trying to connect to Remote EndPoint.

Using Version `4.3.3.952` works fine.

### Which component is your bug related to?
- Client

### To Reproduce
1. Using version of MQTTnet `v4.3.4.1084`.
2. Try to simply invoke Client.ConnectAsync(options)

### Expected behavior
The client connects to the Broker.

### Additional context / logging

```batch
MQTTnet.Exceptions.MqttCommunicationException: None of the discovered or specified addresses match the socket address family. ---> System.ArgumentException: None of the discovered or specified addresses match the socket address family.
  at System.Net.Sockets.SocketAsyncResult.CheckIfThrowDelayedException () [0x00014] in <a8a996a78a804d888710c9e2575d78c8>:0 
  at System.Net.Sockets.Socket.EndConnect (System.IAsyncResult asyncResult) [0x0002c] in <a8a996a78a804d888710c9e2575d78c8>:0 
  at System.Threading.Tasks.TaskFactory`1[TResult].FromAsyncCoreLogic (System.IAsyncResult iar, System.Func`2[T,TResult] endFunction, System.Action`1[T] endAction, System.Threading.Tasks.Task`1[TResult] promise, System.Boolean requiresSynchronization) [0x00019] in <d636f104d58046fd9b195699bcb1a744>:0 
--- End of stack trace from previous location where exception was thrown ---

  at MQTTnet.Implementations.CrossPlatformSocket.ConnectAsync (System.Net.EndPoint endPoint, System.Threading.CancellationToken cancellationToken) [0x000e9] in <b93c2f176a9c4aa59d4db05b60c0955f>:0 
  at MQTTnet.Implementations.MqttTcpChannel.ConnectAsync (System.Threading.CancellationToken cancellationToken) [0x00383] in <b93c2f176a9c4aa59d4db05b60c0955f>:0 
  at MQTTnet.Adapter.MqttChannelAdapter+<>c__DisplayClass30_0.<ConnectAsync>b__1 () [0x0009a] in <b93c2f176a9c4aa59d4db05b60c0955f>:0 
  at MQTTnet.Adapter.MqttChannelAdapter.ConnectAsync (System.Threading.CancellationToken cancellationToken) [0x001e1] in <b93c2f176a9c4aa59d4db05b60c0955f>:0 
   --- End of inner exception stack trace ---
  at MQTTnet.Adapter.MqttChannelAdapter.WrapAndThrowException (System.Exception exception) [0x00090] in <b93c2f176a9c4aa59d4db05b60c0955f>:0 
  at MQTTnet.Adapter.MqttChannelAdapter.ConnectAsync (System.Threading.CancellationToken cancellationToken) [0x00213] in <b93c2f176a9c4aa59d4db05b60c0955f>:0 
  at MQTTnet.Client.MqttClient.ConnectInternal (MQTTnet.Adapter.IMqttChannelAdapter channelAdapter, System.Threading.CancellationToken cancellationToken) [0x000f4] in <b93c2f176a9c4aa59d4db05b60c0955f>:0 
  at MQTTnet.Client.MqttClient.ConnectAsync (MQTTnet.Client.MqttClientOptions options, System.Threading.CancellationToken cancellationToken) [0x002ba] in <b93c2f176a9c4aa59d4db05b60c0955f>:0 
  at MQTTnet.Client.MqttClient.ConnectAsync (MQTTnet.Client.MqttClientOptions options, System.Threading.CancellationToken cancellationToken) [0x0059d] in <b93c2f176a9c4aa59d4db05b60c0955f>:0 

```

### Note
No code provided as the issue is simply related to the ConnectAsync method.

**Comment by chkr1011** [comment]
@daframax Are you using an IP in the _WithTcpServer_ method or an IP address (as string)?

**Comment by daframax** [comment]
Hello @chkr1011,
First of all thank you for the library: it is a great work.
I started using it from version 3.x

Regarding you question: I pass the host as string and the port as int.

**Comment by chkr1011** [comment]
@daframax Please try this version from mygetfeed and let me know if it works: https://www.myget.org/feed/mqttnet/package/nuget/MQTTnet/4.3.4.1116

**Comment by daframax** [comment]
Thank you @chkr1011 I will try it and let you know as soon as possibile

**Comment by chkr1011** [comment]
I figured out that the issue is related to Mono. It handles the _DnsEndpoint_ differently than .NET Framework does. When using the overload with host and port it works perfectly.

Build: [4.3.4.1133](https://www.myget.org/feed/mqttnet/package/nuget/MQTTnet/4.3.4.1133) contains the fix if you want to test.

I will add a workaround for that and release a new version soon.

**Comment by daframax** [comment]
Hello @chkr1011 Thank you for your time.
I was about to report the same thing after testing the version 4.3.4.1116.
I noticed that one of differences between version 4.3.3.952 and 4.3.4.1084 was the ctor into the class CrossPlatformSocket. So I check the source code of the implementation of the class Socket.cs inside the mono repo on github and I noticed that there are differences:

Method *socket.ConnectAsync(string host, int port)* uses:

```csharp
public IAsyncResult BeginConnect (string host, int port, AsyncCallback callback, object state)
		{
			ThrowIfDisposedAndClosed ();

			if (host == null)
				throw new ArgumentNullException ("host");
			if (addressFamily != AddressFamily.InterNetwork && addressFamily != AddressFamily.InterNetworkV6)
				throw new NotSupportedException ("This method is valid only for sockets in the InterNetwork and InterNetworkV6 families");
			if (port <= 0 || port > 65535)
				throw new ArgumentOutOfRangeException ("port", "Must be > 0 and < 65536");
			if (is_listening)
				throw new InvalidOperationException ();

			var sockares = new SocketAsyncResult (this, callback, state, SocketOperation.Connect) {
				Port = port
			};

			var dnsRequest = Dns.GetHostAddressesAsync (host);
			dnsRequest.ContinueWith (t => {
				if (t.IsFaulted)
					sockares.Complete (t.Exception.InnerException);
				else if (t.IsCanceled)
					sockares.Complete (new OperationCanceledException ());
				else {
					sockares.Addresses = t.Result;
					BeginMConnect (sockares);
				}
			}, TaskScheduler.Default);

			return sockares;
		}

```

Method *socket.ConnectAsync(EndPoint endpoint)* uses:
```csharp
public IAsyncResult BeginConnect (EndPoint remoteEP, AsyncCallback callback, object state)
		{
			ThrowIfDisposedAndClosed ();

			if (remoteEP == null)
				throw new ArgumentNullException ("remoteEP");
			if (is_listening)
				throw new InvalidOperationException ();

			SocketAsyncResult sockares = new SocketAsyncResult (this, callback, state, SocketOperation.Connect) {
				EndPoint = remoteEP,
			};

			BeginSConnect (sockares);
			return sockares;
		}
```

I will test the version 4.3.4.1133


Thank you

**Comment by chkr1011** [comment]
Please test with the released 4.3.5 version. I experienced the same issue with one of my Android apps. So, I was able to fix it.
